File: ios/onesygnal/Sources/onesygnal/OneSygnalPlugin.swift
import Flutter
import OneSygnalSDK
import UIKit

// Identifies this plugin's calls to the native SDK as coming from the Flutter bridge rather than
// a native iOS app — see OneSygnal.setSdkWrapper(). Keep in sync with
// apps/flutter-sdk/pubspec.yaml's `version` on every release; there's no runtime way to read the
// Dart package's own version from Swift.
private let wrapperLibrary = "onesygnal-flutter"
private let wrapperVersion = "0.4.0"

/// Thin bridge over the native `OneSygnal` iOS SDK — every method here delegates straight
/// through, no business logic lives in this plugin. Mirrors the equivalent
/// apps/flutter-sdk/android Kotlin plugin exactly (same channel names, same method/event
/// contract) so a single Dart-side implementation can serve both platforms.
///
/// No `import OneSygnalSDK` here deliberately: apps/flutter-sdk/ios/onesygnal.podspec compiles
/// this file together with apps/ios-sdk's own source directly into one CocoaPods pod target
/// (rather than depending on a separately-published OneSygnalSDK pod, which doesn't exist yet —
/// see the podspec's comment) — so `OneSygnal` is already in-module here, same as any other type
/// in this compilation unit.
public class OneSygnalPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    // Registered once per event name in onListen, cancelled in onCancel — only listen to native
    // events while Dart is actually subscribed to the EventChannel stream.
    // contract:events:begin — asserted against fixtures/channel/event-channel.json
    private static let eventNames = [
        "ready",
        "survey:shown",
        "survey:completed",
        "survey:dismissed",
        "survey:question_answered",
    ]
    // contract:events:end

    private var eventSink: FlutterEventSink?
    private var registrations: [OneSygnal.Registration] = []

    // contract:methods:begin — asserted against fixtures/channel/method-channel.json
    // `@escaping` on the `FlutterResult` parameter: "identify"/"initialize"'s handlers below
    // pass `result` into a completion closure handed to the native SDK, which invokes it later
    // (after native work finishes) rather than synchronously within the handler body — a
    // non-escaping `result` can't be captured by that outer escaping closure.
    private lazy var handlers: [String: (FlutterMethodCall, @escaping FlutterResult) -> Void] = [
        "initialize": { _, result in
            OneSygnal.shared.initialize { success in result(success) }
        },
        "setApiKey": { call, result in
            guard let apiKey = call.arguments as? String else { result(nil); return }
            OneSygnal.shared.setApiKey(apiKey)
            result(nil)
        },
        "setLocale": { call, result in
            guard let locale = call.arguments as? String else { result(nil); return }
            OneSygnal.shared.setLocale(locale)
            result(nil)
        },
        "track": { call, result in
            guard let args = call.arguments as? [String: Any], let eventName = args["eventName"] as? String else {
                result(false)
                return
            }
            let properties = args["properties"] as? [String: Any?] ?? [:]
            result(OneSygnal.shared.track(eventName, properties: properties))
        },
        "identify": { call, result in
            guard let args = call.arguments as? [String: Any], let userId = args["userId"] as? String else {
                result(false)
                return
            }
            let attributes = args["attributes"] as? [String: Any]
            OneSygnal.shared.identify(userId, attributes: attributes) { success in result(success) }
        },
        "logout": { _, result in OneSygnal.shared.logout { result(nil) } },
        "reset": { _, result in OneSygnal.shared.reset { result(nil) } },
        "setSurveysEnabled": { call, result in
            guard let enabled = call.arguments as? Bool else { result(nil); return }
            OneSygnal.shared.setSurveysEnabled(enabled)
            result(nil)
        },
        "areSurveysEnabled": { _, result in result(OneSygnal.shared.areSurveysEnabled()) },
        "isInitialized": { _, result in result(OneSygnal.shared.isInitialized()) },
        "shutdown": { _, result in OneSygnal.shared.shutdown { result(nil) } },
    ]
    // contract:methods:end

    var supportedMethods: Set<String> { Set(handlers.keys) }

    public static func register(with registrar: FlutterPluginRegistrar) {
        // Must run before any method channel call — Dart can't call initialize() until this
        // registrar exists, so this always precedes native init/device-context capture.
        OneSygnal.shared.setSdkWrapper(library: wrapperLibrary, version: wrapperVersion)
        let methodChannel = FlutterMethodChannel(name: "onesygnal", binaryMessenger: registrar.messenger())
        let instance = OneSygnalPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(name: "onesygnal_events", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)

        // `detachFromEngineForRegistrar:` below is only called if the plugin instance was
        // published — its own header doc says so. Without this, hot restart accumulates
        // registrations on the process-wide OneSygnal singleton forever.
        registrar.publish(instance)
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        registrations.forEach { $0.cancel() }
        registrations.removeAll()
        eventSink = nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let handler = handlers[call.method] else {
            result(FlutterMethodNotImplemented)
            return
        }
        handler(call, result)
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        for eventName in Self.eventNames {
            let registration = OneSygnal.shared.on(eventName) { [weak self] data in
                var payload: [String: Any] = ["event_type": eventName]
                if let dict = data as? [String: Any] {
                    for (key, value) in dict { payload[key] = value }
                } else if let data {
                    payload["surveyId"] = data
                }
                self?.eventSink?(payload)
            }
            registrations.append(registration)
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        registrations.forEach { $0.cancel() }
        registrations.removeAll()
        eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/app/onesygnal/sdk/flutter/OneSygnalPlugin.kt
package app.onesygnal.sdk.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import app.onesygnal.sdk.api.OneSygnal

// Identifies this plugin's calls to the native SDK as coming from the Flutter bridge rather than
// a native Android app — see OneSygnal.setSdkWrapper(). Keep in sync with
// apps/flutter-sdk/pubspec.yaml's `version` on every release; there's no runtime way to read the
// Dart package's own version from Kotlin.
private const val WRAPPER_LIBRARY = "onesygnal-flutter"
private const val WRAPPER_VERSION = "0.4.0"

/**
 * Thin bridge over the native [OneSygnal] Android SDK — every method here delegates straight
 * through, no business logic lives in this plugin. Mirrors the equivalent
 * apps/flutter-sdk/ios/onesygnal Swift plugin exactly (same channel names, same method/event
 * contract) so a single Dart-side implementation can serve both platforms.
 */
class OneSygnalPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var applicationContext: android.content.Context
    private var activityBinding: ActivityPluginBinding? = null
    private var eventSink: EventChannel.EventSink? = null

    // Registered once per event name in onListen, cancelled in onCancel — only listen to native
    // events while Dart is actually subscribed to the EventChannel stream.
    // contract:events:begin — asserted against fixtures/channel/event-channel.json
    private val eventNames = listOf(
        "ready",
        "survey:shown",
        "survey:completed",
        "survey:dismissed",
        "survey:question_answered",
    )
    // contract:events:end
    private val registrations = mutableListOf<OneSygnal.Registration>()

    // contract:methods:begin — asserted against fixtures/channel/method-channel.json
    @Suppress("UNCHECKED_CAST")
    private val handlers: Map<String, (MethodCall, Result) -> Unit> = mapOf(
        "initialize" to { _, result ->
            // Falls back to applicationContext if no Activity is attached yet — initialize()
            // still works, it just won't have an Activity seeded for CurrentActivityTracker
            // until one resumes (matches native's own "seed(context as? Activity)" fallback).
            OneSygnal.initialize(activityBinding?.activity ?: applicationContext) { success ->
                result.success(success)
            }
        },
        "setApiKey" to { call, result ->
            OneSygnal.setApiKey(call.arguments as String)
            result.success(null)
        },
        "setLocale" to { call, result ->
            OneSygnal.setLocale(call.arguments as String)
            result.success(null)
        },
        "track" to { call, result ->
            val args = call.arguments as Map<String, Any?>
            val eventName = args["eventName"] as String
            val properties = (args["properties"] as? Map<String, Any?>) ?: emptyMap()
            result.success(OneSygnal.track(eventName, properties))
        },
        "identify" to { call, result ->
            val args = call.arguments as Map<String, Any?>
            val userId = args["userId"] as String
            val attributes = args["attributes"] as? Map<String, Any?>
            OneSygnal.identify(userId, attributes) { success -> result.success(success) }
        },
        "logout" to { _, result -> OneSygnal.logout { result.success(null) } },
        "reset" to { _, result -> OneSygnal.reset { result.success(null) } },
        "setSurveysEnabled" to { call, result ->
            OneSygnal.setSurveysEnabled(call.arguments as Boolean)
            result.success(null)
        },
        "areSurveysEnabled" to { _, result -> result.success(OneSygnal.areSurveysEnabled()) },
        "isInitialized" to { _, result -> result.success(OneSygnal.isInitialized()) },
        "shutdown" to { _, result -> OneSygnal.shutdown { result.success(null) } },
    )
    // contract:methods:end

    internal val supportedMethods: Set<String> get() = handlers.keys

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        // Must run before any method channel call — Dart can't call initialize() until this
        // binding exists, so this always precedes native init/device-context capture.
        OneSygnal.setSdkWrapper(WRAPPER_LIBRARY, WRAPPER_VERSION)
        methodChannel = MethodChannel(binding.binaryMessenger, "onesygnal")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "onesygnal_events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        // Without this, a hot restart re-registers the plugin (and re-listens) against the same
        // process-wide OneSygnal singleton without ever releasing the previous engine's listeners.
        registrations.forEach { it.cancel() }
        registrations.clear()
        eventSink = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activityBinding = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activityBinding = binding
    }

    override fun onDetachedFromActivity() {
        activityBinding = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val handler = handlers[call.method] ?: return result.notImplemented()
        handler(call, result)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        for (eventName in eventNames) {
            val listener: (Any?) -> Unit = { data ->
                val payload = mutableMapOf<String, Any?>("event_type" to eventName)
                if (data is Map<*, *>) {
                    @Suppress("UNCHECKED_CAST")
                    payload.putAll(data as Map<String, Any?>)
                } else if (data != null) {
                    payload["surveyId"] = data
                }
                eventSink?.success(payload)
            }
            registrations.add(OneSygnal.on(eventName, listener))
        }
    }

    override fun onCancel(arguments: Any?) {
        registrations.forEach { it.cancel() }
        registrations.clear()
        eventSink = null
    }
}
